llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp

```diff
@@ -996,6 +1006,7 @@ static void simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo) {
                          m_LogicalAnd(m_VPValue(X1), m_Not(m_VPValue(Y1))))) &&
       X == X1 && Y == Y1) {
     R.getVPSingleValue()->replaceAllUsesWith(X);
+    R.eraseFromParent();
```

Comment:
Is this related to this patch or independent?
There are multiple RAUW's here, w/o such erasure, is this one special?

Reply:
It's required.  In my view VPlanTransforms is being lazy with its cleanup as proven here because this simplification is crucial to how predicates are built up for tail-folded loops that employ if-conversion.  Without this change you get none of the expected benefit of the new vblend transformation because the mask's use counts are incorrect.

I'd like to follow up to see if "fixing" the other transforms results in further optimisations, but this was the minimum to solve my initial need.

Reply:
Probably best to wait until https://github.com/llvm/llvm-project/pull/93998 lands (work-list for simplifyRecipes), eager removal may remove recipes that may have been added to worklist earlier.

Once it has been switched to a work-list, we should be able to simplify, clean-up dead recipes and re-simplify the instructions that have any of their users removed.

As the other patterns at the moment aren't guarded by a use-count check, I don't think there should be differences at the moment from cleaning up early.

Reply:
Unless there's a very strong reason I'd rather not wait because tail folded loops are being artificially hampered (as shown via several performance degradations when SVE is enabled) and the unit-tests show only benefits to this PR as it stand today.

Reply:
I don't think this needs to hold up the patch, just want to avoid updating additional places preemptively.